File: 2816.double-a-number-represented-as-a-linked-list.py
#
# @lc app=leetcode id=2816 lang=python3
#
# [2816] Double a Number Represented as a Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def reverse(node):
            prev = None
            curr = node
            while curr:
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt
            return prev

        # Step 1: Reverse the list
        head = reverse(head)
        carry = 0
        curr = head
        prev = None
        # Step 2: Double each digit and handle carry
        while curr:
            total = curr.val * 2 + carry
            curr.val = total % 10
            carry = total // 10
            prev = curr
            curr = curr.next
        # Step 3: If there's a leftover carry, add new node
        if carry:
            prev.next = ListNode(carry)
        # Step 4: Reverse back
        return reverse(head)
# @lc code=end

File: 2816.double-a-number-represented-as-a-linked-list.java
#
# @lc app=leetcode id=2816 lang=java
#
# [2816] Double a Number Represented as a Linked List
#
# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
    public ListNode doubleIt(ListNode head) {
        head = reverse(head);
        ListNode curr = head;
        int carry = 0;
        ListNode prev = null;
        while (curr != null) {
            int sum = curr.val * 2 + carry;
            curr.val = sum % 10;
            carry = sum / 10;
            prev = curr;
            curr = curr.next;
        }
        if (carry > 0) {
            prev.next = new ListNode(carry);
        }
        head = reverse(head);
        // General consistency check: traverse and ensure each node value is 0-9 and structure is intact
        ListNode node = head;
        while (node != null) {
            if (node.val < 0 || node.val > 9) throw new IllegalStateException("Invalid node value");
            node = node.next;
        }
        return head;
    }
    private ListNode reverse(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
}
# @lc code=end